### PR TITLE
added support for H2 database storage

### PR DESCRIPTION
**Pull request by adejanovski**
Add incremental repair column for H2
Bump up version to 0.4.0
Code design fixes

Fetched, rebased and fixed conflicts from @spodkowinski PR on the Spotify repo (thanks!)

### CODE REVIEW

**Comment by spodkowinski** [comment]
Glad to see that this could end up in production after all. 
I've added some commits to my own [h2_storage](https://github.com/spodkowinski/cassandra-reaper-tlp/commits/h2_storage) branch while trying to get those new shiny acceptance tests working with H2. They are really fast, so why not running them with both in-memory and H2 as storage? But I'm still undecided how to pass the two different yaml configs to `ReaperTestJettyRunner`.

**Comment by adejanovski** [comment]
@spodkowinski : I've been thinking about it too and we need to test all storage backends in the acceptance tests.
Picking different yaml files could be done by modifying BasicSteps : 

```
@Given("^a real reaper service is running$")
  public void a_real_reaper_service_is_running() throws Throwable {
    setupReaperIntegrationTestRunner();
```

to : 

```
@Given("^a real reaper service is running with \"([^\"]*)\" as config file$")
  public void a_real_reaper_service_is_running(String configFile) throws Throwable {
    setupReaperIntegrationTestRunner(configFile);
```

Change the call to `ReaperTestJettyRunner.setup(context)` with `ReaperTestJettyRunner.setup(context, configFile)`. 

Then change ReaperTestJetty : 

```
public static void setup(AppContext testContext, String configFile ) throws Exception {
    if (runnerInstance == null) {
      String testConfigPath = Resources.getResource(configFile).getPath();
```

Specifying the config file would then be possible in the Cucumber feature file.

What do you think ?

**Comment by spodkowinski** [comment]
I'm new to cucumber, but spending some time on looking into how the feature spec is supposed to work, I feel like running the same scenarios with different backgrounds isn't something that can be done easily (see https://github.com/cucumber/gherkin/issues/56). Running the same scenarios against different config files (and thus backends) would require to either use something like [scenario outlines](https://github.com/cucumber/cucumber/wiki/Scenario-Outlines) and example data tables for the config file value, or to copy the feature file for each config. But maybe I'm still missing something here?

File: pom.xml [hunk 1]
```diff
@@ -255,10 +263,12 @@
                               <goal>verify</goal>
                             </goals>
                             <configuration>
+                            <forkCount>1</forkCount>
+        					<reuseForks>false</reuseForks>
```
Comment (adejanovski) [inline pom.xml]: MINOR Tab characters found (this is the first occurrence) [rule](https://sonarqube.com/coding_rules#rule_key=xml%3AIllegalTabCheck)

**Comment by adejanovski** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by adejanovski** [comment]
@spodkowinski : I've pushed code to run the acceptance tests against memory and H2, one after the other. 

If you check the commits, you'll see I've removed starting the dropwizard server from the Background section in the Cucumber feature file, and instead use a @BeforeClass Junit annotation to decide which configuration file to use.

Memory test :
```
@RunWith(Cucumber.class)
@CucumberOptions(
    features = "classpath:com.spotify.reaper.acceptance/integration_reaper_functionality.feature"
)
public class ReaperIT {
  private static final Logger LOG = LoggerFactory.getLogger(ReaperIT.class);
  private static ReaperTestJettyRunner runnerInstance;
  private static final String MEMORY_CONFIG_FILE="cassandra-reaper-at.yaml";
  
  
  @BeforeClass
  public static void setUp() throws Exception {
    LOG.info("setting up testing Reaper runner with {} seed hosts defined and memory storage",
        TestContext.TEST_CLUSTER_SEED_HOSTS.size());
    AppContext context = new AppContext();    
    runnerInstance = ReaperTestJettyRunner.setup(context, MEMORY_CONFIG_FILE);

    BasicSteps.setReaperClient(ReaperTestJettyRunner.getClient());
  }
```

H2 test : 
```
@RunWith(Cucumber.class)
@CucumberOptions(
    features = "classpath:com.spotify.reaper.acceptance/integration_reaper_functionality.feature"
)
public class ReaperH2IT {
  private static final Logger LOG = LoggerFactory.getLogger(ReaperH2IT.class);
  private static ReaperTestJettyRunner runnerInstance;
  private static final String H2_CONFIG_FILE="cassandra-reaper-h2-at.yaml";
  
  
  @BeforeClass
  public static void setUp() throws Exception {
    LOG.info("setting up testing Reaper runner with {} seed hosts defined and H2 storage",
        TestContext.TEST_CLUSTER_SEED_HOSTS.size());
    AppContext context = new AppContext();    
    runnerInstance = ReaperTestJettyRunner.setup(context, H2_CONFIG_FILE);

    BasicSteps.setReaperClient(ReaperTestJettyRunner.getClient());
  }
```

This way we can reuse the same feature file but use different backends.
I had to change the failsafe config so that each test runs in a different VM so that it works fine (some stuff seems to be initialized only once in the VM).

**Comment by adejanovski** [comment]
SonarQube analysis reported 8 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [ReaperApplication.java#L224](https://github.com/thelastpickle/cassandra-reaper/blob/059d7fdcabc0db1c0f93400ac9bb096e85390110/src/main/java/com/spotify/reaper/ReaperApplication.java#L224): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L225](https://github.com/thelastpickle/cassandra-reaper/blob/059d7fdcabc0db1c0f93400ac9bb096e85390110/src/main/java/com/spotify/reaper/ReaperApplication.java#L225): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L226](https://github.com/thelastpickle/cassandra-reaper/blob/059d7fdcabc0db1c0f93400ac9bb096e85390110/src/main/java/com/spotify/reaper/ReaperApplication.java#L226): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L227](https://github.com/thelastpickle/cassandra-reaper/blob/059d7fdcabc0db1c0f93400ac9bb096e85390110/src/main/java/com/spotify/reaper/ReaperApplication.java#L227): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L228](https://github.com/thelastpickle/cassandra-reaper/blob/059d7fdcabc0db1c0f93400ac9bb096e85390110/src/main/java/com/spotify/reaper/ReaperApplication.java#L228): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L229](https://github.com/thelastpickle/cassandra-reaper/blob/059d7fdcabc0db1c0f93400ac9bb096e85390110/src/main/java/com/spotify/reaper/ReaperApplication.java#L229): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L230](https://github.com/thelastpickle/cassandra-reaper/blob/059d7fdcabc0db1c0f93400ac9bb096e85390110/src/main/java/com/spotify/reaper/ReaperApplication.java#L230): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [PostgresStorage.java#L221](https://github.com/thelastpickle/cassandra-reaper/blob/059d7fdcabc0db1c0f93400ac9bb096e85390110/src/main/java/com/spotify/reaper/storage/PostgresStorage.java#L221): Remove this unused private "tryDeletingRepairUnit" method. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

**Comment by adejanovski** [comment]
SonarQube analysis reported 8 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [ReaperApplication.java#L224](https://github.com/thelastpickle/cassandra-reaper/blob/6519536d625560ecfa64b01611cf49d84278cd16/src/main/java/com/spotify/reaper/ReaperApplication.java#L224): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L225](https://github.com/thelastpickle/cassandra-reaper/blob/6519536d625560ecfa64b01611cf49d84278cd16/src/main/java/com/spotify/reaper/ReaperApplication.java#L225): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L226](https://github.com/thelastpickle/cassandra-reaper/blob/6519536d625560ecfa64b01611cf49d84278cd16/src/main/java/com/spotify/reaper/ReaperApplication.java#L226): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L227](https://github.com/thelastpickle/cassandra-reaper/blob/6519536d625560ecfa64b01611cf49d84278cd16/src/main/java/com/spotify/reaper/ReaperApplication.java#L227): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L228](https://github.com/thelastpickle/cassandra-reaper/blob/6519536d625560ecfa64b01611cf49d84278cd16/src/main/java/com/spotify/reaper/ReaperApplication.java#L228): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L229](https://github.com/thelastpickle/cassandra-reaper/blob/6519536d625560ecfa64b01611cf49d84278cd16/src/main/java/com/spotify/reaper/ReaperApplication.java#L229): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L230](https://github.com/thelastpickle/cassandra-reaper/blob/6519536d625560ecfa64b01611cf49d84278cd16/src/main/java/com/spotify/reaper/ReaperApplication.java#L230): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [PostgresStorage.java#L221](https://github.com/thelastpickle/cassandra-reaper/blob/6519536d625560ecfa64b01611cf49d84278cd16/src/main/java/com/spotify/reaper/storage/PostgresStorage.java#L221): Remove this unused private "tryDeletingRepairUnit" method. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

**Comment by adejanovski** [comment]
SonarQube analysis reported 8 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [ReaperApplication.java#L224](https://github.com/thelastpickle/cassandra-reaper/blob/8e06277f9496403a9c83195eaa83b9794b55a89b/src/main/java/com/spotify/reaper/ReaperApplication.java#L224): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L225](https://github.com/thelastpickle/cassandra-reaper/blob/8e06277f9496403a9c83195eaa83b9794b55a89b/src/main/java/com/spotify/reaper/ReaperApplication.java#L225): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L226](https://github.com/thelastpickle/cassandra-reaper/blob/8e06277f9496403a9c83195eaa83b9794b55a89b/src/main/java/com/spotify/reaper/ReaperApplication.java#L226): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L227](https://github.com/thelastpickle/cassandra-reaper/blob/8e06277f9496403a9c83195eaa83b9794b55a89b/src/main/java/com/spotify/reaper/ReaperApplication.java#L227): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L228](https://github.com/thelastpickle/cassandra-reaper/blob/8e06277f9496403a9c83195eaa83b9794b55a89b/src/main/java/com/spotify/reaper/ReaperApplication.java#L228): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L229](https://github.com/thelastpickle/cassandra-reaper/blob/8e06277f9496403a9c83195eaa83b9794b55a89b/src/main/java/com/spotify/reaper/ReaperApplication.java#L229): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [ReaperApplication.java#L230](https://github.com/thelastpickle/cassandra-reaper/blob/8e06277f9496403a9c83195eaa83b9794b55a89b/src/main/java/com/spotify/reaper/ReaperApplication.java#L230): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. MAJOR [PostgresStorage.java#L221](https://github.com/thelastpickle/cassandra-reaper/blob/8e06277f9496403a9c83195eaa83b9794b55a89b/src/main/java/com/spotify/reaper/storage/PostgresStorage.java#L221): Remove this unused private "tryDeletingRepairUnit" method. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

**Comment by adejanovski** [comment]
SonarQube analysis reported 8 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ReaperApplication.java#L224](https://github.com/thelastpickle/cassandra-reaper/blob/c408faf48edc798833803d45fbacbf479f18fafc/src/main/java/com/spotify/reaper/ReaperApplication.java#L224): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. ![MAJOR][MAJOR] [ReaperApplication.java#L225](https://github.com/thelastpickle/cassandra-reaper/blob/c408faf48edc798833803d45fbacbf479f18fafc/src/main/java/com/spotify/reaper/ReaperApplication.java#L225): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. ![MAJOR][MAJOR] [ReaperApplication.java#L226](https://github.com/thelastpickle/cassandra-reaper/blob/c408faf48edc798833803d45fbacbf479f18fafc/src/main/java/com/spotify/reaper/ReaperApplication.java#L226): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. ![MAJOR][MAJOR] [ReaperApplication.java#L227](https://github.com/thelastpickle/cassandra-reaper/blob/c408faf48edc798833803d45fbacbf479f18fafc/src/main/java/com/spotify/reaper/ReaperApplication.java#L227): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. ![MAJOR][MAJOR] [ReaperApplication.java#L228](https://github.com/thelastpickle/cassandra-reaper/blob/c408faf48edc798833803d45fbacbf479f18fafc/src/main/java/com/spotify/reaper/ReaperApplication.java#L228): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. ![MAJOR][MAJOR] [ReaperApplication.java#L229](https://github.com/thelastpickle/cassandra-reaper/blob/c408faf48edc798833803d45fbacbf479f18fafc/src/main/java/com/spotify/reaper/ReaperApplication.java#L229): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. ![MAJOR][MAJOR] [ReaperApplication.java#L230](https://github.com/thelastpickle/cassandra-reaper/blob/c408faf48edc798833803d45fbacbf479f18fafc/src/main/java/com/spotify/reaper/ReaperApplication.java#L230): Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2629)
1. ![MAJOR][MAJOR] [PostgresStorage.java#L221](https://github.com/thelastpickle/cassandra-reaper/blob/c408faf48edc798833803d45fbacbf479f18fafc/src/main/java/com/spotify/reaper/storage/PostgresStorage.java#L221): Remove this unused private "tryDeletingRepairUnit" method. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Pull request merged**